checker/src/main/java/org/checkerframework/checker/determinism/DeterminismAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,990 @@
+                return super.visitMethodInvocation(node, annotatedRetType);
+            }
+
+            ExecutableElement m = methodFromUse(node).methodType.getElement();
+
+            // If return type (non-array, non-collection, and non-iterator) resolves to
+            // @OrderNonDet, replaces the annotation on the return type with @NonDet.
+            if (annotatedRetType.getAnnotations().contains(ORDERNONDET)
+                    && !mayBeOrderNonDet(annotatedRetType.getUnderlyingType())) {
+                annotatedRetType.replaceAnnotation(NONDET);
+            }
+
+            // Annotates the return type of "equals()" method called on a Set receiver
+            // as described in the specification of this method.
+
+            // Example1: @OrderNonDet Set<@OrderNonDet List<@Det Integer>> s1;
+            //           @OrderNonDet Set<@OrderNonDet List<@Det Integer>> s2;
+            // s1.equals(s2) is @NonDet
+
+            // Example 2: @OrderNonDet Set<@Det List<@Det Integer>> s1;
+            //            @OrderNonDet Set<@Det List<@Det Integer>> s2;
+            // s1.equals(s2) is @Det
+            // TODO-rashmi: this can be more precise (@Det receiver and @OrderNonDet argument)
+            TypeElement receiverUnderlyingType =
+                    TypesUtils.getTypeElement(receiverType.getUnderlyingType());
+
+            // Without this check, NullPointerException in Collections class with buildJdk.
+            // Likely cause: Collections has a private constructor?
+            // Error at line: public class Collections {
+            // TODO-rashmi: check why?
+            if (receiverUnderlyingType == null) {
+                return super.visitMethodInvocation(node, annotatedRetType);
+            }
+
+            if (isEqualsMethod(m)) {
+                AnnotatedTypeMirror argument = getAnnotatedType(node.getArguments().get(0));
+                if (isSet(receiverUnderlyingType.asType())
+                        && receiverType.hasAnnotation(ORDERNONDET)
+                        && !hasOrderNonDetListAsTypeArgument(receiverType)
+                        && isSet(TypesUtils.getTypeElement(argument.getUnderlyingType()).asType())
+                        && argument.hasAnnotation(ORDERNONDET)
+                        && !hasOrderNonDetListAsTypeArgument(argument)) {
+                    annotatedRetType.replaceAnnotation(DET);
+                }
+            }
+
+            // The following code fixes Issue#14
+            // (https://github.com/t-rasmud/checker-framework/issues/14).
+            // Checks if the return type is not a TYPEVAR, and if the invoked method belongs
+            // to a set of (hardcoded) Collection methods in the JDK that return a generic type.
+            // If the check succeeds, annotates the return type depending on the
+            // type of the receiver and the method invoked.
+            // Note: Annotating a generic type with @PolyDet (or any annotation for that matter)
```

Comment:
I don't understand what the problem with constraining both bounds is.  Can you give a simple test case that would fail without this code?  The test case in the issue uses an enhanced for loop.